src/core/util/data_version.pl

```diff
@@ -0,0 +1,180 @@
+        data_version:read_data_version_header_(Header, Data_Version),
+        error(bad_data_version(Header), _)).
+read_data_version_header(_Request, no_data_version).
+
+read_data_version_header_(Header, data_version(Label, Value)) :-
+    once(sub_atom(Header, Label_Length, 1, Value_Length, ':')),
+    % The lengths should be > 0, but we know they will be larger.
+    Label_Length > 3,
+    Value_Length > 3,
+    once(sub_atom(Header, 0, Label_Length, _After_Label, Label)),
+    Before_Value is Label_Length + 1,
+    once(sub_atom(Header, Before_Value, Value_Length, 0, Value)),
+    % Check for extraneous colons.
+    \+ sub_atom(Value, _, _, _, ':').
+
+:- begin_tests(read_data_version_header_tests).
+
+test("empty header", [fail]) :-
+    read_data_version_header_('', _).
+
+test("empty label and value", [fail]) :-
+    read_data_version_header_(':', _).
+
+test("empty label", [fail]) :-
+    read_data_version_header_(':value', _).
+
+test("empty value", [fail]) :-
+    read_data_version_header_('label:', _).
+
+test("extraneous colons", [fail]) :-
+    read_data_version_header_('label:value1:value2', _).
+
+test("short label", [fail]) :-
+    read_data_version_header_('labe:value', _).
+
+test("short value", [fail]) :-
+    read_data_version_header_('label:valu', _).
+
+test("pass") :-
+    read_data_version_header_('label:value', data_version(label, value)).
+
+:- end_tests(read_data_version_header_tests).
+
+write_data_version_header(data_version(Data_Version_Label, Data_Version_Value)) :-
+    atom(Data_Version_Label),
+    atom(Data_Version_Value),
+    !,
+    format("TerminusDB-Data-Version: ~s:~s~n", [Data_Version_Label, Data_Version_Value]).
+write_data_version_header(no_data_version) :-
+    !.
+write_data_version_header(Data_Version) :-
+    throw(error(unexpected_argument_instantiation(write_data_version_header, Data_Version), _)).
+
+/**
+ * transaction_data_version(+Object, -Data_Version) is semidet.
+ *
+ * Look for the data version of a transaction/validation object. Fail if Object
+ * does not have a data version.
+ */
+transaction_data_version(Object, Data_Version) :-
+    utils:do_or_die(
+        (   is_dict(Object),
+            (   transaction_object{} :< Object
+            ;   validation_object{} :< Object
+            )),
+        error(unexpected_argument_instantiation('transaction_data_version/2', Object), _)),
+    % Check if Object has a parent repository object and set that.
+    (   get_dict(parent, Object, Parent)
+    ->  Repo_Object = Parent
+    ;   Repo_Object = Object
+    ),
+    transaction_data_version_(Object.descriptor, Repo_Object, Data_Version).
+
+/**
+ * transaction_data_version(+Object, +Objects, -Data_Version) is semidet.
+ *
+ * Look for the data version of a transaction/validation object using Objects to
+ * look for Object's repository transaction/validation object. Fail if Object
+ * does not have a data version.
+ */
+transaction_data_version(Object, Objects, Data_Version) :-
```

Comment:
It looks like this case is used for extracting the data version from validation objects only, as for transaction objects we can always just use the parent transaction.
Maybe that can be clarified in a comment.

Reply:
Good point. Maybe we should even go further?

1. Rename `transaction_data_version/3` to `validation_data_version/3`
2. Remove unnecessary checks:
   * `transaction_object{}` from `validation_data_version/3`
   * `validation_object{}` from `transaction_data_version/2`

What do you think?

Reply:
Yeah, that's a good idea. We won't be able to accidentally use it in a wrong way then.

Reply:
https://github.com/terminusdb/terminusdb/pull/911/commits/dc28a52379efcf8b9926961dcbddabc6dd1e6120